### PR TITLE
Updating split validator to allow duplicates across test sets

### PR DESCRIPTION
**Pull request by Andrewq11**
## Changelogs

- Updating the `split` field validator in the `BenchmarkSpecification` model to check for duplicates _within_ individual test sets rather than _across_ all test sets
- Added test to ensure duplicates in a nested test set are caught
- Added test to ensure that duplicate indices are allowed across indices

---

_Checklist:_

- [x] _Was this PR discussed in an issue? It is recommended to first discuss a new feature into a GitHub issue before opening a PR._ https://github.com/polaris-hub/polaris/issues/157
- [x] _Add tests to cover the fixed bug(s) or the newly introduced feature(s) (if appropriate)._
- [x] _Update the API documentation if a new function is added, or an existing one is deleted._
- [x] _Write concise and explanatory changelogs above._
- [x] _If possible, assign one of the following labels to the PR: `feature`, `fix`, `chore`, `documentation` or `test` (or ask a maintainer to do it for you)._

---

_discussion related to that PR_

### CODE REVIEW

**Review by zhu0619 — APPROVED** [review]
Thanks @Andrewq11
Looks good to me.

File: polaris/benchmark/_base.py [hunk 1]
```diff
@@ -182,30 +182,44 @@ def _validate_split(cls, v, info: ValidationInfo):
             raise InvalidBenchmarkError("The predefined split contains empty test partitions")
 
         train_idx_list = v[0]
-        test_idx_list = list(i for part in v[1].values() for i in part) if isinstance(v[1], dict) else v[1]
+        full_test_idx_list = (
+            list(i for part in v[1].values() for i in part) if isinstance(v[1], dict) else v[1]
+        )
```
Comment (mercuryseries) [inline polaris/benchmark/_base.py:187]: While not directly related to the main focus of this PR, I noticed we could potentially optimize this line by using [`itertools.chain.from_iterable`](https://docs.python.org/3/library/itertools.html#itertools.chain.from_iterable):

```suggestion
        full_test_idx_list = list(chain.from_iterable(v[1].values())) if isinstance(v[1], dict) else v[1]
```

File: polaris/benchmark/_base.py [hunk 2]
```diff
@@ -182,30 +182,44 @@ def _validate_split(cls, v, info: ValidationInfo):
...
-        if len(test_idx_set) != len(test_idx_list):
-            raise InvalidBenchmarkError("The test set contains duplicate indices")
+
+        # Check for duplicate indices within a given test set. Because a user can specify
+        # multiple test sets for a given benchmark and it is acceptable for indices to be shared
+        # across test sets, we check for duplicates in each test set independently.
+        if isinstance(v[1], dict):
+            for test_set_name, test_set_idx_list in v[1].items():
+                test_set_idx_set = set(test_set_idx_list)
+                if len(test_set_idx_list) != len(test_set_idx_set):
```
Comment (mercuryseries) [inline polaris/benchmark/_base.py:211]:
```suggestion
                if len(set(test_set_idx_list)) != len(test_set_idx_list):
```

File: polaris/benchmark/_base.py [hunk 3]
```diff
@@ -182,30 +182,44 @@ def _validate_split(cls, v, info: ValidationInfo):
...
+                    raise InvalidBenchmarkError(
+                        f'Test set with name "{test_set_name}" contains duplicate indices'
+                    )
+        else:
+            if len(full_test_idx_set) != len(full_test_idx_list):
+                raise InvalidBenchmarkError("The test set contains duplicate indices")
```
Comment (mercuryseries) [inline polaris/benchmark/_base.py:217]: You can keep it as is if you find the above version more readable:

```suggestion
        elif len(full_test_idx_set) != len(full_test_idx_list):
            raise InvalidBenchmarkError("The test set contains duplicate indices")
```

**Review by mercuryseries — APPROVED** [review]
Thanks @Andrewq11! I added some suggestions.

**Comment by Andrewq11** [comment]
Thanks for the suggestions @mercuryseries!

**Pull request merged**